### PR TITLE
Adds dim kwarg support for cross entropy loss

### PR DESCRIPTION
**Pull request by yashika-git**
### <samp>🤖 Generated by Copilot at dc1331f</samp>

This pull request adds a new `dim` parameter to the `cross_entropy` loss function and its module counterpart, which allows computing the loss along a different dimension of the input. It also updates the documentation and the tests to reflect the new functionality.

Next iteration of  https://github.com/pytorch/pytorch/pull/81397

Fixes #77256 

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @mcarilli @ptrblck @leslie-fang-intel @EikanWang @voznesenskym @penguinwu @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

### CODE REVIEW

**Comment by malfet** [comment]
@yashika-git should we reopen https://github.com/pytorch/pytorch/pull/81397 and continue conversation there?

**Review by malfet — APPROVED** [review]

**Comment by yashika-git** [comment]
> @yashika-git should we reopen #81397 and continue conversation there?

@malfet, Sure, the branch in #81397 is different, shall I make the new changes in that branch then?

**Comment by malfet** [comment]
> @malfet, Sure, the branch in #81397 is different, shall I make the new changes in that branch then?

No need to duplicate the work, let's just reference old PR from this one, so that folks can jog their memories of the previous conversations

**Comment by malfet** [comment]
I guess you'll also need to make a minor tweaks to `functorch`/`torch.func` to keep CI happy

**Comment by yashika-git** [comment]
> I guess you'll also need to make a minor tweaks to `functorch`/`torch.func` to keep CI happy

Alright, I'll do so

File: test/test_nn.py [hunk 1]
```diff
@@ -7019,6 +7019,16 @@ def test_cross_entropy_loss_zero_div(self):
         input_2 = torch.rand([5, 0], dtype=torch.float32)
         torch.nn.CrossEntropyLoss()(input_1, input_2)
 
+    def test_cross_entropy_dim_parameter(self):
```
Comment (jbschlosser) [inline test/test_nn.py:7022]: We really want this within the cross-device tests to make sure we're covering CUDA as well ([link](https://github.com/pytorch/pytorch/blob/5210d7c4233f115d57a09cfe7d96c2a32a057f7c/test/test_nn.py#L11575) to location after the other device-specific cross entropy tests).

File: torch/nn/functional.py [hunk 2]
```diff
@@ -3043,10 +3047,13 @@ def cross_entropy(
...
             reduction=reduction,
             label_smoothing=label_smoothing,
+            dim=dim,
         )
     if size_average is not None or reduce is not None:
         reduction = _Reduction.legacy_get_string(size_average, reduce)
-    return torch._C._nn.cross_entropy_loss(input, target, weight, _Reduction.get_enum(reduction), ignore_index, label_smoothing)
+        if dim is not None:
+            input = input.swapaxes(dim, 1)
+    return torch._C._nn.cross_entropy_loss(input, target, weight, _Reduction.get_enum(reduction), ignore_index, label_smoothing, dim)
```
Comment (jbschlosser) [inline torch/nn/functional.py:3056]: Passing `dim` here isn't going to work. This PR doesn't change the interface of the underlying ATen op exposed via `torch._C._nn.cross_entropy_loss` (and I don't think it should).

Rather, we'll need to swap axes back for the output when we're not doing a mean or sum reduction.

File: test/test_nn.py (hunk 1 above)
Comment (jbschlosser) [reply test/test_nn.py:7022]: Also, we need to make sure that the tests cover all three reduction types. In particular, I don't think `reduction='none'` will work in the current state.

You can follow what was done here to parametrize the test accordingly:
https://github.com/pytorch/pytorch/blob/5210d7c4233f115d57a09cfe7d96c2a32a057f7c/test/test_nn.py#L11555-L11556

File: torch/nn/modules/loss.py [hunk 3]
```diff
@@ -1160,20 +1163,22 @@ class probabilities only when a single class label per minibatch item is too res
...
     """
-    __constants__ = ['ignore_index', 'reduction', 'label_smoothing']
+    __constants__ = ['ignore_index', 'reduction', 'label_smoothing', 'dim']
     ignore_index: int
     label_smoothing: float
+    dim: int
 
     def __init__(self, weight: Optional[Tensor] = None, size_average=None, ignore_index: int = -100,
-                 reduce=None, reduction: str = 'mean', label_smoothing: float = 0.0) -> None:
+                 reduce=None, reduction: str = 'mean', label_smoothing: float = 0.0, dim: int = None) -> None:
```
Comment (jbschlosser) [inline torch/nn/modules/loss.py:1172]: as indicated by the linter, the type here should be `dim: Optional[int]`

**Review by jbschlosser — CHANGES_REQUESTED** [review]
Thanks for the contribution! Just a couple testing and implementation related things and we can get this in

File: torch/nn/modules/loss.py (hunk 3 above)
Comment (yashika-git) [reply torch/nn/modules/loss.py]: @jbschlosser, I've corrected this.

File: torch/nn/functional.py (hunk 2 above)
Comment (yashika-git) [reply torch/nn/functional.py]: @jbschlosser, I've removed `dim` from here, but I couldn't understand how to swap the axes back when we're not doing a mean or sum reduction.

File: test/test_nn.py (hunk 1 above)
Comment (yashika-git) [reply test/test_nn.py]: @jbschlosser, I've tried parameterizing the tests to cover the 3 reduction types, and tried to cover CUDA as well.

File: torch/nn/functional.py (hunk 2 above)
Comment (jbschlosser) [reply torch/nn/functional.py]: something like this should work:
```python
output = torch._C._nn.cross_entropy_loss(input, target, weight, _Reduction.get_enum(reduction), ignore_index, label_smoothing, dim)
if reduction == 'none':
    output = output.swapaxes(dim, 1)
return output
```

this is untested but the tests you wrote should verify it :)

File: torch/nn/modules/loss.py [hunk 4]
```diff
@@ -1160,20 +1163,22 @@ class probabilities only when a single class label per minibatch item is too res
         >>> output = loss(input, target)
         >>> output.backward()
     """
-    __constants__ = ['ignore_index', 'reduction', 'label_smoothing']
+    __constants__ = ['ignore_index', 'reduction', 'label_smoothing', 'dim']
     ignore_index: int
     label_smoothing: float
+    dim: int
```
Comment (jbschlosser) [inline torch/nn/modules/loss.py:1169]: Need `Optional[int]` here as well to satisfy the linter

**Review by jbschlosser — COMMENTED** [review]
Pretty close, just gotta get linter checks / tests passing :) Thanks!

File: test/test_nn.py [hunk 5]
```diff
@@ -7019,6 +7019,17 @@ def test_cross_entropy_loss_zero_div(self):
         input_2 = torch.rand([5, 0], dtype=torch.float32)
         torch.nn.CrossEntropyLoss()(input_1, input_2)
 
+    @parametrize_test("reduction", ("none", "mean", "sum"))
+    def test_cross_entropy_dim_parameter(self, device, reduction):
```
Comment (jbschlosser) [inline test/test_nn.py:7023]: This test should be moved underneath the `TestNNDeviceType` class so it is automatically instantiated across devices. [Here](https://github.com/pytorch/pytorch/blob/5210d7c4233f115d57a09cfe7d96c2a32a057f7c/test/test_nn.py#L11575) is a good place to put it

**Review by jbschlosser — COMMENTED** [review]
Note: I think we'll need to tweak the `cross_entropy` entry in `torch/overrides.py` to include `dim` as well

File: torch/nn/modules/loss.py (hunk 4 above)
Comment (yashika-git) [reply torch/nn/modules/loss.py]: Done

File: test/test_nn.py (hunk 5 above)
Comment (yashika-git) [reply test/test_nn.py]: I've moved the test

File: torch/nn/functional.py (hunk 2 above)
Comment (yashika-git) [reply torch/nn/functional.py]: Thanks, I've added this

**Comment by yashika-git** [comment]
> Note: I think we'll need to tweak the `cross_entropy` entry in `torch/overrides.py` to include `dim` as well

I've included `dim` in `torch/overrides.py` now.

File: test/test_nn.py [hunk 6]
```diff
@@ -11573,6 +11573,17 @@ def test_cross_entropy_large_tensor(self, device, reduction):
...
+        loss = nn.CrossEntropyLoss(reduction=reduction)
+        input_1 = torch.randn(15, 10, 20, dtype=torch.float, device=device)
+        input_2 = input_1.swapaxes(2, 1)
+        target = torch.empty(15, 20, dtype=torch.long, device=device).random_(10)
+
+        loss_1 = loss(input_1, target)
+        loss_2 = loss(input_2, target, dim=2)
```
Comment (jbschlosser) [inline test/test_nn.py:11584]: `dim` should be passed during construction of the `nn.CrossEntropyLoss()` object instead of during the forward pass, no?

Comment (yashika-git) [reply test/test_nn.py]: @jbschlosser, Apologies, my bad, I've corrected it now.

**Comment by yashika-git** [comment]
@jbschlosser, could you please review it? Thanks.

**Comment by mikaylagawarecki** [comment]
Please also fix the failing CI for `test_cross_entropy_dim_parameter`, let me know if you need help with that.

**Comment by yashika-git** [comment]
@mikaylagawarecki, Thank you, I'll try working on fixing the issues over the weekend.

**Comment by yashika-git** [comment]
@mikaylagawarecki, could you please review this? Thank you

File: torch/nn/modules/loss.py [hunk 7]
```diff
@@ -1165,20 +1168,22 @@ class probabilities only when a single class label per minibatch item is too res
...
         super().__init__(weight, size_average, reduce, reduction)
         self.ignore_index = ignore_index
         self.label_smoothing = label_smoothing
+        self.dim = dim
 
     def forward(self, input: Tensor, target: Tensor) -> Tensor:
         return F.cross_entropy(input, target, weight=self.weight,
                                ignore_index=self.ignore_index, reduction=self.reduction,
-                               label_smoothing=self.label_smoothing)
+                               label_smoothing=self.label_smoothing, dim=self.dim)
```
[the remark this answers is not in the document]
Comment (yashika-git) [reply torch/nn/modules/loss.py:1186]: I've tried doing this

File: torch/nn/functional.py [hunk 8]
```diff
@@ -3055,10 +3059,16 @@ def cross_entropy(
             reduce=reduce,
             reduction=reduction,
             label_smoothing=label_smoothing,
+            dim=dim,
         )
     if size_average is not None or reduce is not None:
         reduction = _Reduction.legacy_get_string(size_average, reduce)
-    return torch._C._nn.cross_entropy_loss(input, target, weight, _Reduction.get_enum(reduction), ignore_index, label_smoothing)
+        if dim is not None:
+            input = input.swapaxes(dim, 1)
```
[the remark this answers is not in the document]
Comment (yashika-git) [reply torch/nn/functional.py]: Tried adding this

File: test/test_nn.py [hunk 9]
```diff
@@ -11888,7 +11888,19 @@ def test_cross_entropy_large_tensor(self, device, reduction):
...
-
+            
+    @parametrize_test("reduction", ("none", "mean", "sum"))
+    def test_cross_entropy_dim_parameter(self, device, reduction):
+        loss_1 = nn.CrossEntropyLoss(reduction=reduction)
+        loss_2 = nn.CrossEntropyLoss(reduction=reduction, dim=2)
+        input_1 = torch.randn(15, 10, 20, dtype=torch.float, device=device)
+        input_2 = input_1.swapaxes(2, 1)
+        target = torch.empty(15, 20, dtype=torch.long, device=device).random_(10)
+        if reduction != "none":
```
Comment (mikaylagawarecki) [inline test/test_nn.py:11899]: could we add testing for the reduction = "none" case as well

**Review by mikaylagawarecki — COMMENTED** [review]
@yashika-git this looks ok but it looks like lint is still failing (see https://hud.pytorch.org/pr/pytorch/pytorch/98594#15840211413)

To fix the failing lint job running `setup_lint` and `quicklint` from your local pytorch directory and following the corresponding instructions should be sufficient, see [here](https://github.com/pytorch/pytorch/blob/4721553431e968ee5e90753346d92252cce98e84/Makefile#L25-L34)

File: torch/nn/modules/loss.py [hunk 10]
```diff
@@ -1131,6 +1131,9 @@ class probabilities only when a single class label per minibatch item is too res
             of smoothing when computing the loss, where 0.0 means no smoothing. The targets
             become a mixture of the original ground truth and a uniform distribution as described in
             `Rethinking the Inception Architecture for Computer Vision <https://arxiv.org/abs/1512.00567>`__. Default: :math:`0.0`.
+        dim (int, optional): Specifies the dimension of the classes (the dimension across which softmax is computed).
+            If :attr:`dim` is set to ``None``, then :attr:`dim` defaults to 1.
```
Comment (mikaylagawarecki) [inline torch/nn/modules/loss.py:1135]: could you update this docstring to mention that it is not used for the no-batch-dim case as well`(C)`

**Comment by yashika-git** [comment]
> @yashika-git this looks ok but it looks like lint is still failing (see https://hud.pytorch.org/pr/pytorch/pytorch/98594#15840211413)
> 
> To fix the failing lint job running `setup_lint` and `quicklint` from your local pytorch directory and following the corresponding instructions should be sufficient, see [here](https://github.com/pytorch/pytorch/blob/4721553431e968ee5e90753346d92252cce98e84/Makefile#L25-L34)

@mikaylagawarecki, I ran `make setup_lint` and `make quicklint` commands in my local pytorch directory and the output contains quite a lot of errors (attaching the partial screenshot of the output). Could you please tell if I'm running the correct commands? Thank you.
image

File: test/test_nn.py (hunk 9 above)
Comment (yashika-git) [reply test/test_nn.py:11898]: Tried adding this

**Comment by mikaylagawarecki** [comment]
lint is still failing, see [here for logs](https://github.com/pytorch/pytorch/actions/runs/5858115245/job/15881876691). The commands seem correct, perhaps you need to rebase your PR? Also, did you run `lintrunner -a` after `make quicklint`?

In general, you can go to this link [hud.pytorch.org/pr/98594](https://hud.pytorch.org/pr/98594) to see the failures/logs. There are some new failures for the additions in the most recent commit to `test_cross_entropy_dim_parameter`

**Comment by yashika-git** [comment]
> Hey, lint is still failing, see [here for logs](https://github.com/pytorch/pytorch/actions/runs/5858115245/job/15881876691). The commands seem correct, perhaps you need to rebase? Also, did you run `lintrunner -a` after `make quicklint`?
> 
> In general, you can go to this link [hud.pytorch.org/pr/98594](https://hud.pytorch.org/pr/98594) to see the failures/logs! There are some new failures for the additions in the most recent commit to `test_cross_entropy_dim_parameter`

@mikaylagawarecki, In the most recent commit, I did run `lintrunner -a` but it's still failing.
Could you please help me with rebase? On which branch do I have to rebase over?
Also, I've tried fixing the error in `test_cross_entropy_dim_parameter`.
Thank you
